src/lib/FlightTasks/tasks/AutoMapper2/FlightTaskAutoMapper2.cpp

```diff
@@ -195,3 +195,34 @@ bool FlightTaskAutoMapper2::_highEnoughForLandingGear()
 	// return true if altitude is above two meters
 	return _alt_above_ground > 2.0f;
 }
+
+float FlightTaskAutoMapper2::_getLandSpeed()
+{
+	bool rc_assist_enabled = _param_mpc_land_rc_help.get();
+	bool rc_is_valid = !_sub_vehicle_status->get().rc_signal_lost;
+
+	float throttle = 0.5f;
+
+	if (rc_is_valid && rc_assist_enabled) {
+		throttle = _sub_manual_control_setpoint->get().z;
+	}
+
+	float speed = 0;
+
+	if (_alt_above_ground > _param_mpc_land_alt1.get()) {
+		speed = _constraints.speed_down;
+
+	} else {
+		float land_speed = _param_mpc_land_speed.get();
+		float head_room = _constraints.speed_down - land_speed;
+
+		speed = land_speed + 2 * (0.5f - throttle) * head_room;
```

Comment:
I don't want to offend at all but this implementation is quite hacky and needs improvement for sure.
Issues that come to my mind:
- Stick subscription and evaluation is duplicate to https://github.com/PX4/Firmware/blob/e25db0162019e5f0bb7171ec6f3db2f456672d95/src/lib/FlightTasks/tasks/Manual/FlightTaskManual.cpp#L70-L115 but misses important checks and e.g. deadzone.
- It's implemented in AutoMapper2 which duplicates big parts of AutoMapper so in some configurations this feature is not available.
- We should support general correction of the land location by sticks not only adjusting the land speed.

Reply:
No offense taken at all, I like critical feedback. I am pretty unfamiliar with FlightTasks in general, so indeed it may be hacky as I was unsure the correct spot to put this. 

> We should support general correction of the land location by sticks not only adjusting the land speed.

I've got a WIP branch open right now. My plan was to move the sticks stuff into the `FlightTask` base class so I could use `_evaluateSticks`

> It's implemented in AutoMapper2 which duplicates big parts of AutoMapper so in some configurations this feature is not available.

I was very confused by the naming of these two. Without getting intimate with the code, it is very unclear what their roles/responsibilities are. I actually implemented it `AutoMapper` at first and then realized nothing was happening. 

If you've got any suggestions or just random details in general don't hesitate to share. Thanks!

Reply:
> I've got a WIP branch open right now. My plan was to move the sticks stuff into the FlightTask base class so I could use _evaluateSticks

I'm quite sure by now that inheriting all the data processing in the flight tasks does not scale. We should pull e.g. the stick processing into a library such that not every task that uses sticks has to inherit from Manual. (The same of course also for other reusable logic.)

> I was very confused by the naming of these two.

That's definitely far from your fault. They do almost completely the same thing but for later auto task with the "AutoSmoothVel" ending (which is default) there are some small differences. We need to get rid of that duplication, here's the issue for that: https://github.com/PX4/Firmware/issues/11209. The existence of Automapper and having two versions is in my eyes partly caused by the inheritance structure.

Reply:
@dakejahl JFYI the duplicate auto mapper is gone since https://github.com/PX4/Firmware/pull/13596/commits/7e78eed58642833d27795a13fad39547c382bf35 🎉

Reply:
@dakejahl JFYI the stick handling was moved to a library (#15324) and I have a pr open reusing it for the auto land use case you contributed: https://github.com/PX4/Firmware/pull/15325

Reply:
Sweet! Have others found it useful? Haha I never did get around to allowing XY control, it would be cool if it did

Reply:
I think it doesn't get used that often because of RC override which by default makes it switch out of land when you move the sticks. We'll enable x, y control soon. We can still discuss how it should work together with the mode override.